### ISSUE TITLE
Automatically integrations to initiate based on Chrome Group Policy Object

### ISSUE THREAD

**Issue by twschiller** [pixiebrix/pixiebrix-extension#2869]
Context
- IT teams provision PixieBrix via Chrome Group Policy Object
- These are set via https://chromeenterprise.google/policies/?policy=ExtensionInstallForcelist
- They pre-allow permissions to pages/API endpoints using: runtime_allowed_hosts
- In the force_install scenario, we want PixieBrix to be able to detect which context it's running in (e.g., which company is using it), to apply some co-branding and also simplify connecting to enterprise-specific services

To verify:
- Are permission from runtime_allowed_hosts available from https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/permissions/getAll

Related:
- https://support.google.com/chrome/a/answer/7532015?hl=en#zippy=%2Cset-installation-policies-automatically-install-force-install-allow-or-block%2Cprevent-apps-and-extensions-from-altering-webpages
- [There's no API to read the policy from the extension](https://support.google.com/chrome/a/thread/4663625/how-to-read-the-extensionpolicy-from-chrome-extension?hl=en)
- Slack discussion: https://pixiebrix.slack.com/archives/D024R1S2D9N/p1646227301874669

**Comment by fregante** [comment]
On a second read, the only mention of `runtime_allowed_hosts` comes attached to the `runtime_blocked_hosts` key, exclusively as an exception to the latter:

> `runtime_allowed_hosts` — Allows extensions to interact with specified websites, even if they’re also defined in runtime_blocked_hosts.

and it's in the section:

> Prevent apps and extensions from altering webpages

I don’t think it can be used on its own and I think it won't automatically grant access to those hosts even if it could be.

> * we want PixieBrix to be able to detect which context it's running in

As you mentioned, this isn't an API, so I don’t think we can access it.

> * Are permission from runtime_allowed_hosts available from [developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/permissions/getAll](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/permissions/getAll)

What I said would mean that that key _does not_ grant permissions, so the user would still have to manually grant permission to such hosts.



> * In the force_install scenario, we want PixieBrix to be able to detect which context it's running in (e.g., which company is using it), to apply some co-branding and also simplify connecting to enterprise-specific services

It doesn't look like Chrome tells us about _why_ it was installed in the `onInstalled` callback:

> https://developer.chrome.com/docs/extensions/reference/runtime/#type-OnInstalledReason

On that whole `runtime`  API page there's a single, unrelated mention to the enterprise policy, so I don’t think there's any way to even ask the installation origin of the extension.


---

In short, I don't see anything that we can use to facilitate the use from these users in any way. I suppose what you can do would need to come exclusively via the account, e.g.:

1. User logs in with `dude@corp.net`
2. Account comes preloaded with extensions
3. OnBoarding has `requestPermission` as one of the required steps

**Comment by twschiller** [comment]
Continuing conversation from Slack. I think the analysis above is correct. Sam from our team also gave things a try on their Windows machine with Group Policy Editor

I opened a w3c working group ticket here: https://github.com/w3c/webextensions/issues/174

Open Questions
- Currently when using runtime_allowed_hosts, the Page Editor still prompts the user to to grant permissions. But when the user clicks, they do not have to accepts permissions from a Chrome popup. The dev tools is using calling [canAccessTab](http://github.com/pixiebrix/pixiebrix-extension/blob/main/src/pageEditor/context.ts#L88-L88). What does this us under the hood? Trying to inject code? Calling https://developer.chrome.com/docs/extensions/reference/permissions/#method-contains? I wonder if the Page Editor has access but doesn't realize it? Or if is doesn't, we could request access without it being triggered by a user interaction?
- We had removed the identity.email manifest permissions since we weren't using it. If we re-add it for the workaround described in the linked ticket, will people with normal installs have to reactivate PixieBrix? I believe if people have PixieBrix force_installed they won't have to re-enable (since there's no way to toggle PixieBrix if it's force installed)

**Comment by twschiller** [comment]
According to Chrome Dev Rel, the correct way to solve this is using managed_storage which allows the IT admin to pass data to the extension matching a schema. I will create a separate ticket for layering in support for managed storage

* https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/managed
